### ISSUE TITLE
Check if the second helicity filtering in fortran for mirror processes can be removed

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#941]
Hi @oliviermattelaer , 

as discussed in https://github.com/madgraph5/madgraph4gpu/pull/935#issuecomment-2256328718

This is a reminder to check if the second helicity filtering in fortran for mirror processes can be removed.

For the moment, my proposal to solve #872 via PR #935 is to 
- keep computing twice helicities in fortran
- keep computing only once helicities in cudacpp
- add a second reset_cumulative_variable call in cudacpp for mirri processes (to fix 872 and ensure that frtran and cudacpp give the same results)
- add a sanity check in fortran that the two lists of good helicities are the same

If, as you suggest, we eventually find that helicities can be computed only once in fortran, then
- the second reset_cumative_variable call can be removed in cudacpp
- there is no need for a sanity chec in fortran any more, as helicities are computed only once

I assign this to you... Low priority, in my opinion
Thanks
Andrea

**Comment by oliviermattelaer** [comment]
The fact to have Ntry/goodhel to have a mirror index is ... as old as the [matrix_madevent_group_v4.inc](https://github.com/mg5amcnlo/mg5amcnlo/commit/382e71efd61ef2c2f19be5664b1228220e095ebe#diff-14a4ad6338146e3a56c090d8f0ded955b9291b060f1d9db5e4e636ef2e742090) file (this is even the original reason of such file):
https://github.com/mg5amcnlo/mg5amcnlo/tree/382e71efd61ef2c2f19be5664b1228220e095ebe
(version number is 1.0.0 -> 2011)

Now I have added a test comparison that crash at fortran stage if this is not the case
- [x] p p > t t~ 
- [ ] p p > t t~ j 
- [ ] p p > t t~ j j
- [ ] p p > w+ j
- [ ] p p > w+ j j 
- [ ] p p > w+ j j j
- [ ] p p > w+ w+ j j QCD=0
- [ ] p p > w+ w+ j j QCD=0 , W+ > l+ vl 
- [ ] p p > w+ w- j j 
- [ ] p p > w+ w- j j QCD=0
- [ ] p p > w+ w- j j QCD<=4

**Comment by oliviermattelaer** [comment]
Adding only those helicities to the fortran code does not change (at all) the cross-section (so bit to bit compatibility and really 0 helicity
```
     Cross-section :   11.35 +- 0.03021 pb
     Nb of events :  10000
```
versus
```
     Cross-section :   11.35 +- 0.03021 pb
     Nb of events :  10000
```

**Comment by valassi** [comment]
Hi @oliviermattelaer thanks for looking at this :-)

I was wondering this morning, does Madgraph include the option of polarised beams? Because one physics case where I imagine different helicity lists is that of polarised beams. Well, typically that would be e+ against e- which is not symmetric anyway, but in principle if you have something like e- against e- and the left and right beams have different polarisations, then the list of helicities are different. Far fetched, but maybe something similar is the reason why this was introduced at the time?

(Reminder from yesterday's meeting: you said that the computation of helicities in fortran separately for each mirror was introduced by JohannA in one of the first bug fix releases 1.1.0 or similar, "to fix a bug" or similar, so maybe there was a use case).

Anyway, in the meantime as discussed I will look at introducing a sanity check that if two lists of helicities are computed in fortran, then they are the same.

**Comment by oliviermattelaer** [comment]
So the commit message of Johan is: `Fixed problem when different helicities are zero for main process and mirror process for grouped subprocess output`
related information in the UpdateNotes:
```
                   JA: Improved helicity selection and automatic full helicity
                       sum if needed. Optimization of run parameters.
```

**Comment by valassi** [comment]
> ```
> 
> ```
> 
> 
>     
>   
> 
> So the issue is occuring due to uu>uu that has some helicity that contribute only for that process
> while in itself uu~ and u~u does have symmetric goodhel as expected but goodhel is here a global variable for all processes... Making it non global can create issue for the grid handling (and therefore for loop-induced/gridpack?)
> 
> But the good news is that for "gpu grouping" (so what we have now, this is indeed symmetric).
> This still needs more investigation to answer the question to with if we can handle this without the need of imirror or not.

Hi Olivier, thanks again for looking at this :-)

I implemented my sanity checks, I see no issues in pp->ttjj (well, actually in pp_tt012j).

Now, there is one important difference: in the cudacpp generated code, remember that we split the processes (maybe this is what you mean by gpu grouping?). In particular pp -> tt 0,1,2 jets has
```
ls -d1 pp_tt012j.mad/SubProcesses/P*
pp_tt012j.mad/SubProcesses/P0_gg_ttx/
pp_tt012j.mad/SubProcesses/P0_uux_ttx/
pp_tt012j.mad/SubProcesses/P1_gg_ttxg/
pp_tt012j.mad/SubProcesses/P1_gu_ttxu/
pp_tt012j.mad/SubProcesses/P1_gux_ttxux/
pp_tt012j.mad/SubProcesses/P1_uux_ttxg/
pp_tt012j.mad/SubProcesses/P2_gg_ttxgg/
pp_tt012j.mad/SubProcesses/P2_gg_ttxuux/
pp_tt012j.mad/SubProcesses/P2_gu_ttxgu/
pp_tt012j.mad/SubProcesses/P2_gux_ttxgux/
pp_tt012j.mad/SubProcesses/P2_uc_ttxuc/
pp_tt012j.mad/SubProcesses/P2_ucx_ttxucx/
pp_tt012j.mad/SubProcesses/P2_uu_ttxuu/
pp_tt012j.mad/SubProcesses/P2_uux_ttxccx/
pp_tt012j.mad/SubProcesses/P2_uux_ttxgg/
pp_tt012j.mad/SubProcesses/P2_uux_ttxuux/
pp_tt012j.mad/SubProcesses/P2_uxcx_ttxuxcx/
pp_tt012j.mad/SubProcesses/P2_uxux_ttxuxux/

\grep MIRRORPROCS pp_tt012j.mad/SubProcesses/P* -r | \grep DATA
pp_tt012j.mad/SubProcesses/P0_gg_ttx/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
pp_tt012j.mad/SubProcesses/P0_uux_ttx/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P1_gg_ttxg/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
pp_tt012j.mad/SubProcesses/P1_gu_ttxu/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P1_gux_ttxux/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P1_uux_ttxg/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_gg_ttxgg/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
pp_tt012j.mad/SubProcesses/P2_gg_ttxuux/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
pp_tt012j.mad/SubProcesses/P2_gu_ttxgu/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_gux_ttxgux/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uc_ttxuc/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_ucx_ttxucx/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uu_ttxuu/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
pp_tt012j.mad/SubProcesses/P2_uux_ttxccx/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uux_ttxgg/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uux_ttxuux/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uxcx_ttxuxcx/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.TRUE./
pp_tt012j.mad/SubProcesses/P2_uxux_ttxuxux/mirrorprocs.inc:      DATA (MIRRORPROCS(I),I=1,1)/.FALSE./
```

In particular:
- there is no qq because quark-quark and quark-antiquark collisions are separate (so uu_XXXX and uux_XXXX are separate)
- there are a few directories with MIRRORPROCS=TRUE, I checked only P2_uux_ttxuux and P2_uc_ttxuc and in both cases the helicity lists from the two mirrors match

So, bottomline: remember that in cudacpp generated code (including generated fortran) we do NOT mix uu_XXXX and uux_XXXX together, so if these two have separate helicities, this is handled because they are separate directories. Changing this would require a much more general handling of nprocesses>2, which so far we do not use.

Rephrasing: I think that adding this sanity check should be enough, and in cudacpp we can safely keep two helicity computations in fortran and only one in cudacpp. Do you agree?

**Comment by oliviermattelaer** [comment]
Yes, this is what I meant by GPU grouping.

Now this makes me realize that they are steps that I do not like in the way those helicities are handle and I did create a new branch(https://github.com/mg5amcnlo/mg5amcnlo/tree/goodhel) to implement some change and to allow to experiment more on this without being affected by the issue spotted for tt~.  I do agree  that the above tt~ case is not relevant for this plugin. 

Now I still need to investigate this since
- the commit message of Johan and the status of the code at the time, clearly linked that this is possible for the mirror
- I LIKE to solve puzzle and this is a quite  intriguing/funny one.
- This a good opportunity to clean the mess that goodhel is at madevent level anyway 

The obviously side effect of this fortran refactoring is that your patching method will likely crash since the base file will be different.

**Comment by valassi** [comment]
Hi Olivier, thanks. Yes good to investigate puzzles :-)

Just one question, are you trying to completely remove the second helicity calculation? Or to understand whty it is needed and keep it there? I mean I'd like to understand if eventually we need to implement two helicities in cudacpp

**Comment by oliviermattelaer** [comment]
First step is to understand if /when it is needed.
And if the answer is "it is not needed" -> then we can remove it.

Cheers,

Olivier

On 31 Jul 2024, at 11:39, Andrea Valassi ***@***.***> wrote:



Hi Olivier, thanks. Yes good to investigate puzzles :-)

Just one question, are you trying to completely remove the second helicity calculation? Or to understand whty it is needed and keep it there? I mean I'd like to understand if eventually we need to implement two helicities in cudacpp

**Comment by oliviermattelaer** [comment]
With the new goodhel branch, performing additional tests (first in the SM):
 - [x] p p > t t~ j j
 - [x] p p > t t~ j j j 
 - [x] p p > w+ w+ j j , w+ > l+ vl
 - [x] p p > w+ w+ j j QCD=0 , w+ > l+ vl
 - [x]  p p > w+ w+ j j j QCD=1 , w+ > l+ vl
 - [x] p p > w+ w- j j  QCD<=4 , w+ > l+ vl
 - [x] p p > w+{T} w-{0} j j  QCD<=4 , w+ > l+ vl, w- > j j
 - [x] p p > w+{T} w-{0} j j  QCD=0 , w+ > l+ vl, w- > j j
 - [x] p p > w+{T} w-{0} j j  QCD<=0 

Running test with sm-ckm model (like CMS often use) then the crash was on the first process (that I did test):
 - [ ]  p p > w+{T} w-{0} j j  QCD<=0

So I'm going to investigate  that process now.

**Comment by oliviermattelaer** [comment]
Continue checking for more processes (MSSM):
 - [x] p p > go go
 - [x] p p > go go j j 
 - [x] p p > sq sq j j 
 - [x] (sm) p p > l+ l- j j j
 - [x] (sm) p p > l+ vl j j j

**Comment by valassi** [comment]
> So given that the limit to be "accepted is LIMHEL=1e-8" and that here we are (after 12 events) hitting the first event with only 4 times bigger than the limit... and that many other channel (like G1, does not include 18 for process 63 (in none of the imirror), I think that this is safe to say that this is a threshold effect. So yes here non identical list are occuring and are "fine" so here a simple check of identical list on the fortran side will just lead to an un-necessary crash. (and the result will be equivalent (up to single precision calculation).

Hi Olivier, thanks, nice finds :-)

One point: in the fortran generated through cudacpp, we use LIMHEL=0, since more than two years, see #419. So there is no sensistivity to threshold effect. We had discussed this extensively at the time.

For the release, I would keep this as it is (LIMHEL=0) and add the check that the two helicity lists are the same, to ensure that cudacpp and fortran tests give bit by bit the same results. From a software testing point of view, this is essential.

Later on, this can be rediscussed. One can even think of computing helicities in fortran and passing them to cudacpp.

That said, I still kind of favour having LIMHEL=0 in production full stop. I wonder how much performance we gain and in which specific processes by using LIMHEL>0. But definitely I do not have the full picture here.

**Comment by oliviermattelaer** [comment]
For sm-ckm model, using limhel=0 can giveperformance hit of a factor of 4 or more ...
So this is not negligeable.
Obviously, I do not want to have a "normal" Fortran and a "cudacpp" fortran.
You can set limhel to 0 for test if you want but in production this should be the same fortran code/design.

Cheers,

Olivier





On 2 Aug 2024, at 09:34, Andrea Valassi ***@***.***> wrote:



So given that the limit to be "accepted is LIMHEL=1e-8" and that here we are (after 12 events) hitting the first event with only 4 times bigger than the limit... and that many other channel (like G1, does not include 18 for process 63 (in none of the imirror), I think that this is safe to say that this is a threshold effect. So yes here non identical list are occuring and are "fine" so here a simple check of identical list on the fortran side will just lead to an un-necessary crash. (and the result will be equivalent (up to single precision calculation).

Hi Olivier, thanks, nice finds :-)

One point: in the fortran generated through cudacpp, we use LIMHEL=0, since more than two years, see #419<https://github.com/madgraph5/madgraph4gpu/issues/419>. So there is no sensistivity to threshold effect. We had discussed this extensively at the time.

For the release, I would keep this as it is (LIMHEL=0) and add the check that the two helicity lists are the same, to ensure that cudacpp and fortran tests give bit by bit the same results. From a software testing point of view, this is essential.

Later on, this can be rediscussed. One can even think of computing helicities in fortran and passing them to cudacpp.

That said, I still kind of favour having LIMHEL=0 in production full stop. I wonder how much performance we gain and in which specific processes by using LIMHEL>0. But definitely I do not have the full picture here.

**Comment by oliviermattelaer** [comment]
So I guess that we can close this issue with the conclusion: Yes it can be removed.
This will be done independently of the cudacpp plugin, I will follow the normal procedure for including such fix in 3.6.0.
Which means that this associated issue will automatically be fixed when the merging with 3.6.0 will be complete.
(which is our next big task anyway).

**Comment by valassi** [comment]
> So I guess that we can close this issue with the conclusion: Yes it can be removed.
> This will be done independently of the cudacpp plugin, I will follow the normal procedure for including such fix in 3.6.0

Thanks a lot Olivier :-)

In the meantime while waiting for 360 etc, can we then merge #935 which adds the second reset cumulative call? Then of course this will need to be removed in 360 (I can do that at that moment, no problem), but at least in the meantime while waiting for 360 we have some internally consistent code where fortran and cudacpp are expected to give the same xsec bit by bit. I prefer to add a fix now that we need to remove later, rather than do nothing and have tests that keep failing. Esepecially because I want to test the CMS xsec issues and I need that bit by bit equality.

> You can set limhel to 0 for test if you want but in production this should be the same fortran code/design.

Note, LIMHEL=0 is now in cudacpp production (NOT just in the madgrah4gpu repo). I am ok to remove that and move it ONLY to the madgraph4gpu repo, but can we at least add a runcard that sets limhel=0?  (Or maybe it exists?) I think that would be very useful for debugging.
